Add Intel CET to getTargetInfo

ErnyTech opened the pull request:
This follows PR #15415 which added Intel CET IBT support and LDC PR [#4437](https://github.com/ldc-developers/ldc/pull/4437) to add support for the new CET target in order to maintain a common interface between DMD and LDC.

Later it would be useful to do the same for GDC as well.

SixthDot reviewed (approved):
I suggest to use "intelCET" instead of just "CET" as the later is not very well known.

dkorpel commented:
Why does D code need to know about this setting?

ErnyTech commented:
> Why does D code need to know about this setting?

Mainly for inline ASM

RazvanN7 reviewed (commented):
Please add a changelog entry describing this change so that users get to know about this.

ErnyTech commented:
> Please add a changelog entry describing this change so that users get to know about this.

Added

RazvanN7 commented on changelog/dmd.intel-cet-ibt-protection.dd:
```diff
@@ -0,0 +1,22 @@
+Added support for Intel CET (Control-flow Enforcement Technology) IBT (Indirect Branch Tracking) protection
+
+CET is a technology that is useful for preventing an attacker from redirecting a program's control flow,
+specifically IBT prevents an an attacker from causing an indirect branch to go to an unintended place
```
```suggestion
specifically IBT prevents an attacker from causing an indirect branch to go to an unintended place.
```

Commit:
Add Intel CET to getTargetInfo

This follows PR #15415 which added Intel CET IBT support and
LDC PR #4437 to add support for the new CET target in order
to maintain a common interface between DMD and LDC.

Later it would be useful to do the same for GDC as well.

WalterBright commented on changelog/dmd.intel-cet-ibt-protection.dd:
```diff
@@ -0,0 +1,22 @@
+
+---
+// IBT active
+static assert(__traits(getTargetInfo, "CET") == 1);  // CET == 1 if IBT is active
```
I'm reluctant to call this "CET" when the switch is "IBT". CET includes shadow stack support, which DMD does not yet support. When it does support CET, I'll add a -fCET and only then should getTargetInfo say it supports CET.

ErnyTech replied:
CET includes both IBT and shadow stack, LDC use `__traits(getTargetInfo, "CET") ` as follows:

- Disabled: `__traits(getTargetInfo, "CET") == 0`
- IBT (--fcf-protection=branch in LDC): `__traits(getTargetInfo, "CET") == 1`
- Shadow stack (--fcf-protection=return in LDC): `__traits(getTargetInfo, "CET") == 2`
- IBT + Shadow stack (--fcf-protection=full in LDC): `__traits(getTargetInfo, "CET") == 3`

This behavior is equivalent to `__CET__` of C compilers

WalterBright replied:
ok then

ErnyTech commented:
@dkorpel Is there anything blocking this PR? I think it would be good to include it in 2.106.0-beta especially since it also provides the changelog for -fIBT
Walter's request seems satisfied

ibuclaw commented:
This test is non-portable. Maybe there should be an x86 directory added for such tests (this would include all DMD `asm` tests too).
